### PR TITLE
Binarization

### PR DESCRIPTION
**Pull request by giop98**

### CODE REVIEW

File: server/src/threads.cpp [hunk 1]
```diff
@@ -153,8 +153,8 @@ static void *th_noise_hdl(void *args) {
 
     while(1) {
         read(pctx->noise_r_fd, &show_ready, sizeof show_ready);
-        cv::namedWindow("Noise Visualizer", cv::WINDOW_GUI_NORMAL);
-        cv::imshow("Noise Visualizer", *(show_ready->nframe));
+        cv::namedWindow("Black And White Visualizer", cv::WINDOW_GUI_NORMAL);
```
Comment (MatteoBattilana) [inline server/src/threads.cpp:156]: This is not only used for black and white visualization; better use a more generic name

Comment (giop98) [reply server/src/threads.cpp:156]: Yes, you are right. I changed it during testing, but it's better to change it

Comment (giop98) [reply server/src/threads.cpp]: Changed it to "Modified Image Visualizer"

File: server/src/kernels.cu [hunk 2]
```diff
@@ -248,7 +408,18 @@ diff::cuda::CUDACore::CUDACore(uint8_t *charsPx, matsz& charsSz, float *k, int t
...
+    cudaMalloc((void **)&d_diff, total * sizeof *d_diff);
+    cudaMalloc((void **)&d_xs, total * sizeof *d_xs);
+    cudaMalloc((void **)&d_current, total * sizeof *d_current);
+    cudaMalloc((void **)&d_previous, total * sizeof *d_previous);
+    cudaMalloc((void **)&d_filtered, total * sizeof *d_filtered);
+    cudaMalloc((void **)&d_noise_visualization, total * sizeof *d_noise_visualization);
+    cudaMalloc((void **)&d_grayscale, total * sizeof *d_grayscale);
+    cudaMalloc((void **)&d_binarize, total * sizeof *d_binarize);
+    cudaMalloc((void **)&d_histogram, 256 * sizeof *d_histogram);
+    cudaMalloc((void **)&d_indexes_max, 2 * sizeof *d_indexes_max);
```
Comment (thegabriele97) [inline server/src/kernels.cu:421]: Duplicated mallocs and missing `CUDA_CHECK` macro
```suggestion
    CUDA_CHECK(cudaMalloc((void **)&d_grayscale, total * sizeof *d_grayscale));
    CUDA_CHECK(cudaMalloc((void **)&d_binarize, total * sizeof *d_binarize));
    CUDA_CHECK(cudaMalloc((void **)&d_histogram, 256 * sizeof *d_histogram));
    CUDA_CHECK(cudaMalloc((void **)&d_indexes_max, 2 * sizeof *d_indexes_max));
```

File: server/src/kernels.cu [hunk 3]
```diff
@@ -345,24 +518,88 @@ void diff::cuda::CUDACore::exec_core(uint8_t *frameData, uint8_t *showReadyNData
         kernel2_char<<<1, nThreadsToUse>>>(d_current, d_charsPx + idx * fullArea, eachThreadDoes, offset, 3 * charsSz.width, 3 * frameSz.width);
     }
 
+#ifdef NOISE_VISUALIZER
+#if NOISE_VISUALIZER == 4
+
+    // grayscale_kernel<<<1, nMaxThreads>>>(d_current, d_grayscale, maxAtTime);
+    // grayscale_kernel_v2<<<1, nMaxThreads>>>(d_current, d_grayscale, max4);
+    grayscale_kernel_v3<<<1, nMaxThreads>>>(d_current, d_grayscale, max4);
+    cudaMemcpyAsync(showReadyNData, d_grayscale, total, cudaMemcpyDeviceToHost);
```
Comment (thegabriele97) [inline server/src/kernels.cu:527]:
```suggestion
    CUDA_CHECK(cudaMemcpyAsync(showReadyNData, d_grayscale, total, cudaMemcpyDeviceToHost));
```

File: server/src/kernels.cu [hunk 4]
```diff
@@ -345,24 +518,88 @@ void diff::cuda::CUDACore::exec_core(uint8_t *frameData, uint8_t *showReadyNData
...
+    int index_max = -1, index_sec_max = -1;
+
+    cudaMemcpyAsync(d_histogram, h_histogram, 256 * sizeof(int), cudaMemcpyHostToDevice);
+    grayscale_kernel_v3<<<1, nMaxThreads>>>(d_current, d_grayscale, max4);
+    // generate_histogram<<<1, nMaxThreads>>>(d_grayscale, d_histogram, maxAtTime);
+    generate_histogram_v2<<<1, nMaxThreads>>>(d_grayscale, d_histogram, maxAtTime);
+    compute_max<<<1, 256>>>(d_histogram, d_indexes_max);
+    // cudaMemcpyAsync(showReadyNData, d_grayscale, total, cudaMemcpyDeviceToHost);
+    cudaMemcpyAsync(h_histogram, d_histogram, 256 * sizeof(int), cudaMemcpyDeviceToHost);
+    cudaMemcpyAsync(h_indexes, d_indexes_max, 2 * sizeof(uint8_t), cudaMemcpyDeviceToHost);
```
Comment (thegabriele97) [inline server/src/kernels.cu:541]:
```suggestion
    CUDA_CHECK(cudaMemcpyAsync(d_histogram, h_histogram, 256 * sizeof(int), cudaMemcpyHostToDevice));
    grayscale_kernel_v3<<<1, nMaxThreads>>>(d_current, d_grayscale, max4);
    // generate_histogram<<<1, nMaxThreads>>>(d_grayscale, d_histogram, maxAtTime);
    generate_histogram_v2<<<1, nMaxThreads>>>(d_grayscale, d_histogram, maxAtTime);
    compute_max<<<1, 256>>>(d_histogram, d_indexes_max);
    // cudaMemcpyAsync(showReadyNData, d_grayscale, total, cudaMemcpyDeviceToHost);
    CUDA_CHECK(cudaMemcpyAsync(h_histogram, d_histogram, 256 * sizeof(int), cudaMemcpyDeviceToHost));
    CUDA_CHECK(cudaMemcpyAsync(h_indexes, d_indexes_max, 2 * sizeof(uint8_t), cudaMemcpyDeviceToHost));
```

File: server/src/kernels.cu [hunk 5]
```diff
@@ -345,24 +518,88 @@ void diff::cuda::CUDACore::exec_core(uint8_t *frameData, uint8_t *showReadyNData
...
+    // printf("\nIstogramma\n");
+
+    // for (int i = 0; i < 256; i++) {
+    //     printf("%d\n", h_histogram[i]);
+    // }
+    // printf("\nMassimi\n");
+
+    index_max = h_indexes[0];
+    index_sec_max = h_indexes[1];
+    // // printf("%d %d\n", index_max, index_sec_max);
```
Comment (thegabriele97) [inline server/src/kernels.cu:552]: If the `printf` are useless, you can remove them to have a cleaner code

File: server/src/kernels.cu [hunk 6]
```diff
@@ -345,24 +518,88 @@ void diff::cuda::CUDACore::exec_core(uint8_t *frameData, uint8_t *showReadyNData
...
+    if (threshold < 50) {
+        threshold = 50;
+    }
+    if (threshold > 200) {
+        threshold = 200;
+    }
+
+    // binarize_kernel<<<1, nMaxThreads>>>(d_binarize, d_grayscale, maxAtTime, threshold);
+    binarize_kernel_v2<<<1, nMaxThreads>>>(d_binarize, d_grayscale, max4, threshold);
+    cudaMemcpyAsync(showReadyNData, d_binarize, total, cudaMemcpyDeviceToHost);
```
Comment (thegabriele97) [inline server/src/kernels.cu:563]:
```suggestion
    CUDA_CHECK(cudaMemcpyAsync(showReadyNData, d_binarize, total, cudaMemcpyDeviceToHost));
```

**Review by thegabriele97 — CHANGES_REQUESTED** [review]

File: server/src/kernels.cu [hunk 7]
```diff
@@ -313,27 +473,33 @@ void diff::cuda::CUDACore::exec_core(uint8_t *frameData, uint8_t *showReadyNData
...
-    #ifdef NOISE_VISUALIZER
-        #if NOISE_VISUALIZER == 1         
-        heat_map<<<1, nMaxThreads, 0>>>(d_current, d_previous, max4, d_noise_visualization);
-        CUDA_CHECK(cudaMemcpyAsync(showReadyNData, d_noise_visualization, total, cudaMemcpyDeviceToHost));
-        #elif NOISE_VISUALIZER == 2
-        red_black_map<<<1, nMaxThreads, 0>>>(d_current, d_previous, max4, d_noise_visualization);
-        CUDA_CHECK(cudaMemcpyAsync(showReadyNData, d_noise_visualization, total, cudaMemcpyDeviceToHost));
-        #endif
-    #endif
+#endif
```
Comment (thegabriele97) [inline server/src/kernels.cu:489]: Is this duplicated?

File: server/src/kernels.cu (hunk 5 above)
Comment (giop98) [reply server/src/kernels.cu]: Yes, good suggestion. Now they are removed

File: server/src/kernels.cu (hunk 7 above)
Comment (thegabriele97) [reply server/src/kernels.cu:489]: Let me check

Comment (thegabriele97) [reply server/src/kernels.cu]: Solved https://github.com/MatteoBattilana/CUDAVideoStream/pull/4/commits/28d352251f33b667fb3f3c61601069de427f55c1

**Review by thegabriele97 — APPROVED** [review]

File: server/src/kernels.cu [hunk 8]
```diff
@@ -345,24 +505,80 @@ void diff::cuda::CUDACore::exec_core(uint8_t *frameData, uint8_t *showReadyNData
...
+
+    // grayscale_kernel<<<1, nMaxThreads>>>(d_current, d_grayscale, maxAtTime);
+    // grayscale_kernel_v2<<<1, nMaxThreads>>>(d_current, d_grayscale, max4);
+    grayscale_kernel_v3<<<1, nMaxThreads>>>(d_current, d_grayscale, max4);
+    CUDA_CHECK(cudaMemcpyAsync(showReadyNData, d_grayscale, total, cudaMemcpyDeviceToHost));
+
+#elif NOISE_VISUALIZER == 5
+    int h_histogram[256] = {0};
+    uint8_t h_indexes[2] = {0};
+    int index_max = -1, index_sec_max = -1;
```
Comment (thegabriele97) [inline server/src/kernels.cu:519]: Let me know If I can move them away from here

File: server/src/kernels.cu [hunk 9]
```diff
@@ -345,24 +505,80 @@ void diff::cuda::CUDACore::exec_core(uint8_t *frameData, uint8_t *showReadyNData
...
+    cudaMemcpyAsync(h_xs, d_xs, *h_pos * sizeof *d_xs, cudaMemcpyDeviceToHost);
+    cudaDeviceSynchronize();
+    
+// Noise visualization
+#ifdef NOISE_VISUALIZER
+#if NOISE_VISUALIZER == 3
+    red_black_map_overlap<<<1, nMaxThreads, 0>>>(d_pos, d_xs, (*h_pos) / nMaxThreads, d_previous);
+    CUDA_CHECK(cudaMemcpyAsync(showReadyNData, d_previous, total, cudaMemcpyDeviceToHost));
+#endif
+#endif
```
Comment (thegabriele97) [inline server/src/kernels.cu:578]: I think they are duplicated, needs a check
```suggestion
```

**Review by thegabriele97 — CHANGES_REQUESTED** [review]

File: server/src/kernels.cu (hunk 9 above)
Comment (thegabriele97) [reply server/src/kernels.cu:578]: I am checking

Comment (thegabriele97) [reply server/src/kernels.cu]: Solved https://github.com/MatteoBattilana/CUDAVideoStream/pull/4/commits/81223d16227c30126552d30cb84635bc64f9dc72

File: server/include/common.h [hunk 10]
```diff
@@ -7,6 +7,8 @@
 #define TILE_SIZE 10
 #define BLOCK_SIZE (TILE_SIZE + K - 1)
 
+// Noise visualizer: 1 heatmap, 2 red-black, 3 red-black overlap, 4 grayscale-kernel
+#define NOISE_VISUALIZER 5
```
Comment (MatteoBattilana) [inline server/include/common.h:11]: Why is NOISE_VISUALIZER set to 5?

File: server/include/common.h [hunk 11]
```diff
@@ -7,6 +7,8 @@
...
 // Noise visualizer: 1 heatmap, 2 red-black, 3 red-black overlap
 // #define NOISE_VISUALIZER 1
```
Comment (MatteoBattilana) [inline server/include/common.h:13]: This line is duplicated

**Review by MatteoBattilana — CHANGES_REQUESTED** [review]

File: server/include/common.h (hunk 11 above)
Comment (giop98) [reply server/include/common.h:13]: Yes, I will remove it

File: server/src/kernels.cu [hunk 12]
```diff
@@ -345,24 +505,80 @@ void diff::cuda::CUDACore::exec_core(uint8_t *frameData, uint8_t *showReadyNData
...
+#if NOISE_VISUALIZER == 4
+
+    // grayscale_kernel<<<1, nMaxThreads>>>(d_current, d_grayscale, maxAtTime);
+    // grayscale_kernel_v2<<<1, nMaxThreads>>>(d_current, d_grayscale, max4);
+    grayscale_kernel_v3<<<1, nMaxThreads>>>(d_current, d_grayscale, max4);
+    CUDA_CHECK(cudaMemcpyAsync(showReadyNData, d_grayscale, total, cudaMemcpyDeviceToHost));
+
+#elif NOISE_VISUALIZER == 5
+    int h_histogram[256] = {0};
+    uint8_t h_indexes[2] = {0};
```
[the remark this answers is not in the document]
Comment (giop98) [reply server/src/kernels.cu:518]: By checking the code, I noticed that they are not needed anymore

File: server/src/kernels.cu [hunk 13]
```diff
@@ -345,24 +505,80 @@ void diff::cuda::CUDACore::exec_core(uint8_t *frameData, uint8_t *showReadyNData
...
+    int index_max = -1, index_sec_max = -1;
+
+    CUDA_CHECK(cudaMemcpyAsync(d_histogram, h_histogram, 256 * sizeof(int), cudaMemcpyHostToDevice));
```
[the remark this answers is not in the document]
Comment (giop98) [reply server/src/kernels.cu:521]: Do you think it's better to remove line 517 also if I removed the initialization to 0?

File: server/src/kernels.cu (hunk 8 above)
Comment (giop98) [reply server/src/kernels.cu:519]: I removed index_max and index_sec_max

**Comment by giop98** [comment]
I should have fixed all things and cleaned the code

**Review by thegabriele97 — APPROVED** [review]

**Review by MatteoBattilana — APPROVED** [review]

**Pull request merged**